templates/rhpam73-authoring-ha.yaml

```diff
@@ -318,17 +318,17 @@ parameters:
   required: true
 ## OpenShift Enhancement BEGIN
 - displayName: Prefer KIE Server OpenShift Service [Tech Preview]
-  description: Enables connection to KIE Server via OpenShift internal Service endpoint (Sets the org.kie.server.controller.openshift.prefer.kieserver.service system property)
+  description: If OpenShift Integration of the Business Central is turned on, set this parameter to true to enables connection to KIE Server via an OpenShift internal Service endpoint. Otherwise the Business Central uses a Route to connect to KIE Server. (Sets the org.kie.server.controller.openshift.prefer.kieserver.service system property)
```

Comment:
I do not believe "Otherwise the Business Central uses a Route to connect to KIE Server." is  correct. @rhtevan , please suggest a better description?

Reply:
@errantepiphany Correct me if I'm wrong, the **'org.kie.server.location'** property will to be set as either secure or non-secure KIE Server Route as Template has always defined 'KIE_SERVER_ROUTE_NAME' env. 

https://github.com/rhtevan/jboss-kie-modules/blob/7bbb9a42c739f42cf2fee9c1397445309543492c/jboss-kie-kieserver/added/launch/jboss-kie-kieserver.sh#L346

https://github.com/rhtevan/rhpam-7-openshift-image/blob/c065ee4968de8a106effeb0378f84cddfdf6bdc8/templates/rhpam74-managed.yaml#L967

Reply:
@errantepiphany 
According to the script function, 'org.kie.sever.location' property can be set as non-Route endpoint potentially. Having said that, neither KIE_SERVER_LOCATION nor KIE_SERVER_URL is defined at any existing Template, whereas KIE_SERVER_ROUTE_NAME is being set.

Reply:
@rhtevan , @mramendi : Evan, you are correct, however my point was that org.kie.server.location is *not* used for communication between BC and KIE Servers, no matter if you enable OpenShiftStartupStrategy (OSSS) or not. It is this:
- With OSSS, a ConfigMap is used.
- Without OSS, but with controller using websockets (which is what we do), the org.kie.server.location is _only_ used to create the link shown to users in BC.
- Without OSS, but with controller using REST (which we do _not_ do), _then_ org.kie.server.location is used for communication from BC to KIE Server.

Reply:
@errantepiphany @rhtevan I have removed the part about the Route.

Reply:
@errantepiphany @mramendi 
Based on my observation, I'm still a bit of skeptical of that WebSocket is always the protocol being used for BC and KIE communication. (I may be overlooked something). Here's why.

If we look into these code (See below), which implementations (REST vs WS) selected at runtime is dictated by the ServerTemplate.serverInstanceKey.url that is propagated from 'org.kie.server.location'. Depending on url format, starting with 'http' or 'ws', correspondent implementation will be selected. 

https://github.com/rhtevan/droolsjbpm-integration/blob/c4dba5df4c73f1ab809ebe675f265609723db089/kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/KieServerInstanceManager.java#L506

From OpenShift integration of Business Central perspective, the ServerTemplate is directly converted from KieServerCM/KieServerState. In particular, ServerInstanceKey is constructed by the value from KieServeConfigItem: 'org.kie.server.location'. Therefore, BC should be communicating with KIE Server through REST when OpenShift integration of BC being enabled.

Reply:
@rhtevan @mramendi Well if that's really the case, then each KIE Server should be reporting it's own physical Pod location vs. using the route (or even the Service), as when BC talks to KIE Servers, it needs to talk to each one. If it uses the Route (or Service), then it is load-balanced and there is no guarantee which Pod it is talking to. And in fact, will only talk to just one anyways. I thought this was the whole point of using the OpenShiftStartupStategy... BC communicates ServerTemplate state to KIE Servers via ConfigMaps, and that's it... If I'm wrong, we have a bigger problem.

Reply:
Also, this is why I always complained about needing 2 separate properties. Instead of using one org.kie.server.location system property for both link display in the console, as well as REST interface communication, then it becomes convoluted. There should really be another property... like org.kie.server.location.external or ui or... something.

Reply:
Also, if this is the case, then this issue has been a bug for a looong time now...

Reply:
@errantepiphany 
The reason why BC/EmbeddedController needs to have point-to-point WebSocket connections to every KIE Pod originally is that #1. KIE Pod manages its own local state (aka KieServerState); #2. BC/EmbeddedController is responsible for maintaining a shared state for KIE Server and propagating up-to-date state to every KIE Pod. 

With OpenShift enabled KIE Server (aka OSSS) and BC, Openshift API server manages the shared state for all KIE Pods. All KIE Pods (aka KIE Server instance) do NOT maintain its own local state. The state change can be initiated from BC by invoking KIE Server REST/API, the shared state (aka KieServerCM) will be updated by one of the KIE Pod, which in turn triggers a DC rollout to propagate the state change by refreshing the KIE Pods. As a result, BC/EmbeddedController has no knowledge of individual KIE Pod.